### PR TITLE
[RHBOP] [BXMSPROD-1826] Added nightly jenkins pipeline

### PR DESCRIPTION
**Pull request by lampajr**
### JIRA

- https://issues.redhat.com/browse/BXMSPROD-1826

### Referenced pull requests

- https://github.com/kiegroup/kie-jenkins-scripts/pull/1327

### Checklist
- [ ] Documentation updated if applicable.
- [ ] Release notes updated if applicable.
- [ ] Upgrade recipe provided if applicable.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] tests</b>
  
- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: 
    please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: 
    add the label `run_fdb`

- for a <b>compile downstream build</b>  
  please add comment: <b>Jenkins run cdb</b>

- for a <b>full production downstream build</b>  
  please add comment: <b>Jenkins execute product fdb</b>

- for an <b>upstream build</b>  
  please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] mandrel</b>
</details>

### CI Status

 You can check OptaPlanner repositories CI status from [Chain Status webpage](https://kiegroup.github.io/optaplanner/).

### CODE REVIEW

File: .ci/jenkins/Jenkinsfile.prod.nightly [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            steps {
+                script {
+                    if(env.ALREADY_BUILT_PROJECTS?.trim()) {
+                        echo "[INFO] Start uploading sources zip files.."
+
+                        def PME_BUILD_VARIABLES = env.PME_BUILD_VARIABLES.split(';').collect{ it.split('=')}.inject([:]) {map, item -> map << [(item.length == 2 ? item[0] : null): (item.length == 2 ? item[1] : null)]}
+                        def folder="${env.RCM_GUEST_FOLDER}/rhbop/RHBOP-${PRODUCT_VERSION}.nightly"
+
+                        // optaplanner
+                        echo "[INFO] Start uploading optaplanner sources zip file.."
```
Comment (Ginxo) [inline .ci/jenkins/Jenkinsfile.prod.nightly:86]: I would create function for this and we will have in the `upload` stage one line per source to upload, wdyt?

File: .ci/jenkins/Jenkinsfile.prod.nightly [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                        def PME_BUILD_VARIABLES = env.PME_BUILD_VARIABLES.split(';').collect{ it.split('=')}.inject([:]) {map, item -> map << [(item.length == 2 ? item[0] : null): (item.length == 2 ? item[1] : null)]}
+
+                        def optaplannerArchiveUrl = "https://bxms-qe.rhev-ci-vms.eng.rdu2.redhat.com:8443/nexus/content/groups/rhbop-main-nightly/"
+                        def optaplannerSourcesFileUrl = "${env.STAGING_SERVER_URL}rhbop/RHBOP-${PRODUCT_VERSION}.nightly/rhbop-${PRODUCT_VERSION}-optaplanner-sources-${PME_BUILD_VARIABLES['datetimeSuffix']}.zip"
+                        def optaplannerQuickstartsSourcesFileUrl = "${env.STAGING_SERVER_URL}rhbop/RHBOP-${PRODUCT_VERSION}.nightly/rhbop-${PRODUCT_VERSION}-optaplanner-quickstarts-sources-${PME_BUILD_VARIABLES['datetimeSuffix']}.zip"
+                        def optawebSourcesFileUrl = "${env.STAGING_SERVER_URL}rhbop/RHBOP-${PRODUCT_VERSION}.nightly/rhbop-${PRODUCT_VERSION}-optaweb-vehicle-routing-sources-${PME_BUILD_VARIABLES['datetimeSuffix']}.zip"
+                        def topic = "VirtualTopic.qe.ci.ba.rhbop.${env.UMB_VERSION}.nightly.trigger"
```
Comment (Ginxo) [inline .ci/jenkins/Jenkinsfile.prod.nightly:128]:
```suggestion

                        def topic = "VirtualTopic.qe.ci.ba.rhbop.${env.UMB_VERSION}.nightly.trigger"
```

**Review by Ginxo — CHANGES_REQUESTED** [review]
please review

File: .ci/jenkins/Jenkinsfile.prod.nightly (hunk 1 above)
Comment (lampajr) [reply .ci/jenkins/Jenkinsfile.prod.nightly:86]: I agree, doing it right now

**Comment by lampajr** [comment]
UMB Message looks ok:
```json
{
    "version": {"rhbop":"8.29.0.redhat-2022092908","drools":"8.29.0.redhat-2022092908"},
    "scm_hash": {"drools":"edecaf4a668a1fb931e33d621b86ae021dd1e987","optaplanner":"f2d8f28b34eecbd320f163b842c39b5fbab26ffc","optaplanner-quickstarts":"1b7256ea32447729aa1b94215e68823f60cb6afa","optaweb-vehicle-routing":"40957b766f083bcee3188c7ae60507cf36dd30cb"},
    "built_projects": ["drools","optaplanner","optaplanner-quickstarts","optaweb-vehicle-routing"],
    "archives": "https://bxms-qe.rhev-ci-vms.eng.rdu2.redhat.com:8443/nexus/content/groups/rhbop-main-nightly/",
    "sources": {
        "optaplanner": "http://download.eng.bos.redhat.com/rcm-guest/staging/rhbop/RHBOP-8.29.0.nightly/rhbop-8.29.0-optaplanner-sources-2022092908.zip",
        "optaplanner-quickstarts": "http://download.eng.bos.redhat.com/rcm-guest/staging/rhbop/RHBOP-8.29.0.nightly/rhbop-8.29.0-optaplanner-quickstarts-sources-2022092908.zip",
        "optaweb-vehicle-routing": "http://download.eng.bos.redhat.com/rcm-guest/staging/rhbop/RHBOP-8.29.0.nightly/rhbop-8.29.0-optaweb-vehicle-routing-sources-2022092908.zip"
    }
}
```

File: .ci/jenkins/Jenkinsfile.prod.nightly [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    "archives": "${optaplannerArchiveUrl}",
+    "sources": {
+        "optaplanner": "${optaplannerSourcesUrl}",
+        "optaplanner-quickstarts": "${optaplannerQuickstartsSourcesUrl}",
+        "optaweb-vehicle-routing": "${optawebSourcesFileUrl}"
+    }
+}"""    
+}
+
+def uploadSources(String project, String folder, Map pmeVariables) {
```
Comment (Ginxo) [inline .ci/jenkins/Jenkinsfile.prod.nightly:183]: don't you think datetimeSuffix is just required and not the whole thing? :thinking:

**Review by Ginxo — CHANGES_REQUESTED** [review]
check my comment

File: .ci/jenkins/Jenkinsfile.prod.nightly (hunk 3 above)
Comment (lampajr) [reply .ci/jenkins/Jenkinsfile.prod.nightly:183]: do you mean to have filenames like `optaplanner-quickstarts-sources-2022092909.zip`, i.e., removing the rhbop-8.29.0 prefix?

Comment (Ginxo) [reply .ci/jenkins/Jenkinsfile.prod.nightly:183]: I mean we can probably change method contract to
```
def uploadSources(String project, String folder, String suffix) {
```

Comment (Ginxo) [reply .ci/jenkins/Jenkinsfile.prod.nightly:183]: and then call it by `uploadSources("optaplanner", folder, PME_BUILD_VARIABLES['datetimeSuffix'])`

**Pull request merged**